### ISSUE TITLE
Where is scatter and gather op?

### ISSUE THREAD

**Issue by muazhuda** [webmachinelearning/webnn#467]
Is there reason why it is not in spec?
tensorflow.js supports gather even for webgpu backend.

**Comment by anssiko** [comment]
Thanks for your question. 

We have published guidelines for proposing new ops: https://github.com/webmachinelearning/webnn/blob/main/CONTRIBUTING.md#proposing-and-adding-a-new-operation

If you can answer those questions in this issue, the WG is able to look at your request sooner.

The `gather` op has been already proposed in #375 and the WG is actively discussing it.

**Comment by inexorabletash** [comment]
re: Scatter

* https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update
* https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_.html#torch.Tensor.scatter_
* https://onnx.ai/onnx/operators/onnx__ScatterElements.html
* https://apple.github.io/coremltools/mlmodel/Format/NeuralNetwork.html#scatterlayerparams

Open questions for scatter:
* Behavior if indices contains duplicates? (TF: sums; ONNX: explicit reduction options; PyTorch: undefined, or use scatter_reduce)
* Treatment of axis? (ONNX: explicit; PyTorch: explicit; TF: ??)
* Do all support slice updates like TF?

**Comment by huningxin** [comment]
> re: Scatter

+1 to support scatter, in particular, `scatterND`

```js
MLOperand scatterNd(MLOperand input, MLOperand indices, MLOperand updates);
```

When prototyping Whisper decoders' inference with MLBuffer, I found `scatterND` operator is useful to insert KV values by position into the pre-allocated static KV cache (WebNN requires static shape). 

```js
builder.scatterNd(past_key, position_ids, present_key);
builder.scatterNd(past_value, position_ids, present_value);
```

This avoids reading the KV cache tensor back to CPU and improves the performance. The initial prototype is available at: https://github.com/huningxin/onnxruntime-inference-examples/blob/whisper-mlbuffer/js/whisper-demo/whisper.js

Platforms' support:
- DirectML [DML_SCATTER_ND_OPERATOR_DESC](https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_scatter_nd_operator_desc): Copies the whole input tensor to the output, then overwrites selected indices with corresponding values from the updates tensor.
- CoreML [scatter_nd](https://apple.github.io/coremltools/source/coremltools.converters.mil.mil.ops.defs.html#coremltools.converters.mil.mil.ops.defs.iOS15.scatter_gather.scatter_nd): Scatter `updates` to `data` at locations `indices`. Support "update" and other modes.
- TFLite [scatter_nd](https://www.tensorflow.org/mlir/tfl_ops#tflscatter_nd_tflscatterndop): Scatter sparse `updates` according to individual values at the specified `indices`. Only support "add"/"sum" mode. Need the emulation to support scattering to input tensor.

**Comment by inexorabletash** [comment]
> * TFLite [scatter_nd](https://www.tensorflow.org/mlir/tfl_ops#tflscatter_nd_tflscatterndop): Scatter sparse `updates` according to individual values at the specified `indices`. Only support "add"/"sum" mode. Need the emulation to support scattering to input tensor.

Can't scatter_nd_update be used for the last case?

The impression I got from a quick survey was that "update" into an input tensor was more common in the APIs vs scattering into a new tensor given the shape, but it sounds like models need both?

**Comment by huningxin** [comment]
@inexorabletash 

> Can't TF's [scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update) be used for the last case (scattering to input tensor)?
> 
> The impression I got from a quick survey was that "update" into an input tensor was more common in the APIs vs scattering into a new tensor given the shape, but it sounds like models need both?

I agree "update" into an input tensor is more commonly used, including the updating static KV cache use case and some other models, like [SAM ViT Base](https://huggingface.co/Xenova/sam-vit-base).

TF's [scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update) could be mapped directly. But I am not sure whether it is available in TFLite.

> ... or can we get away with the former only?

+1 to former only. The latter can be emulated by updating into a zero initialized tensor.

**Comment by inexorabletash** [comment]
> TF's [scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update) could be mapped directly. But I am not sure whether it is available in TFLite as backend.

Ooof, yeah. Not listed in https://www.tensorflow.org/mlir/tfl_ops

**Comment by huningxin** [comment]
> > TF's [scatter_nd_update](https://www.tensorflow.org/api_docs/python/tf/tensor_scatter_nd_update) could be mapped directly. But I am not sure whether it is available in TFLite as backend.
> 
> Ooof, yeah. Not listed in https://www.tensorflow.org/mlir/tfl_ops

Although `scatter_nd_update` is not supported as tflite built-in ops, I am not sure whether it could be supported by [Select TensorFlow operators](https://ai.google.dev/edge/litert/models/ops_select) feature. It is listed in the [Supported Select TensorFlow operators](https://ai.google.dev/edge/litert/models/op_select_allowlist).

/cc @reillyeon

**Comment by reillyeon** [comment]
We'll have to see what the binary size impact of adding support for these operators to Chromium's built-in copy of TFLite is, and also what level of support delegates have for them.
